apollo-entity/src/main/java/com/spotify/apollo/entity/EntityMiddleware.java

```diff
@@ -0,0 +1,113 @@
+ * -/-/-
+ */
+package com.spotify.apollo.entity;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.spotify.apollo.Response;
+import com.spotify.apollo.route.AsyncHandler;
+import com.spotify.apollo.route.Middleware;
+import com.spotify.apollo.route.SyncHandler;
+
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+import okio.ByteString;
+
+/**
+ * Apollo {@link Middleware}s for route handlers that work with a typed entity.
+ */
+public interface EntityMiddleware {
+
+  static EntityMiddleware forCodec(EntityCodec codec) {
+    return new CodecEntityMiddleware(codec);
+  }
+
+  static EntityMiddleware forCodec(EntityCodec codec, String contentType) {
+    return new CodecEntityMiddleware(codec, contentType);
+  }
+
+  static EntityMiddleware forJackson(ObjectMapper objectMapper) {
+    return new CodecEntityMiddleware(new JacksonEntityCodec(objectMapper));
+  }
+
+  static EntityMiddleware forJackson(ObjectMapper objectMapper, String contentType) {
+    return new CodecEntityMiddleware(new JacksonEntityCodec(objectMapper), contentType);
+  }
+
+  <E> Middleware<EntityHandler<E, E>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> requestEntityClass);
+
+  <E, R> Middleware<EntityHandler<E, R>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> requestEntityClass, Class<? extends R> responseEntityClass);
+
+  <E> Middleware<EntityResponseHandler<E, E>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> requestEntityClass);
+
+  <E, R> Middleware<EntityResponseHandler<E, R>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> requestEntityClass, Class<? extends R> responseEntityClass);
+
+  <E> Middleware<EntityAsyncHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> requestEntityClass);
+
+  <E, R> Middleware<EntityAsyncHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> requestEntityClass, Class<? extends R> responseEntityClass);
+
+  <E> Middleware<EntityAsyncResponseHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> requestEntityClass);
+
+  <E, R> Middleware<EntityAsyncResponseHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> requestEntityClass, Class<? extends R> responseEntityClass);
+
+  <R> Middleware<SyncHandler<R>, SyncHandler<Response<ByteString>>>
+  serializerDirect(Class<? extends R> responseEntityClass);
+
+  <R> Middleware<SyncHandler<Response<R>>, SyncHandler<Response<ByteString>>>
+  serializerResponse(Class<? extends R> responseEntityClass);
+
+  <R> Middleware<AsyncHandler<R>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerDirect(Class<? extends R> responseEntityClass);
+
+  <R> Middleware<AsyncHandler<Response<R>>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerResponse(Class<? extends R> responseEntityClass);
+
+  interface EntityHandler<E, R>
```

Comment:
These type signatures are hard to decipher (for me). Is it necessary to have them extend ```SyncHandler``` for them to fit in with the rest of Apollo? I've only thought about this rather than tried it, but I get the feeling it shouldn't be, since the middleware will end up translating it into an ```AsyncHandler<Response<ByteString>>```. If it is not necessary, I would prefer to not extend that type, since doing so signals that these are endpoint handlers in their own right. And they're not, they just happen to have the same signature. Or?